File: tkalert/bin/alert.py
# Copyright (C) 2013 NETWAYS GmbH, http://netways.de
#
# This file is part of TKALERT (http://www.thomas-krenn.com/).
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""
    module:: alert
"""

import sys
reload(sys)
sys.setdefaultencoding("utf-8")

import time
import os.path

from datetime import datetime

import tkalert
import logging
from tkalert.options import MyOptions, MyOptionIsMandatoryError
from tkalert.data import HeartbeatObject, AlertObject, \
    map_alert_object_to_arguments
from tkalert.settings import MAIL_SERVER, MAIL_TARGET_ADDRESS, GNUPG_KEY, VERSION_STRING
from tkalert.mail import Mailer
from tkalert.gnupg import GnupgCommand, GnupgCommandException


def main():
    """Main script to trigger alerting
        Returns:
            int. The return code
    """

    time_start = time.time()

    myoptions = MyOptions(usage="%prog --type=<heartbeat|service|test> [--help]",
                          version="%prog " + tkalert.__version__)

    try:
        (options, args) = myoptions.parse_args()

        if options.verbose is True:
            logging.getLogger().setLevel(logging.DEBUG)

        log = logging.getLogger(__name__)
        log.debug('Starting up')

        log.debug('Testing gnupg environment')
        gnupg_paths = [
            options.gnupgconfig,
            '/etc/tkalert/gnupg.conf',
            '/usr/local/tkalert/etc/gnupg.conf'
        ]

        for test_path in gnupg_paths:
            if test_path and os.path.exists(test_path):
                log.debug('setting --gnupg-config=' + test_path)
                options.gnupgconfig = test_path
                break

        xml_object = None

        if options.type == "heartbeat" or options.type == "test" :
            log.info('Creating heartbeat object')
            xml_object = HeartbeatObject()
        elif options.type == "service":
            log.info('Creating alert object')
            xml_object = AlertObject()

        xml_object.set_authkey(options.auth)
        xml_object.set_contact_email(options.mail)
        xml_object.set_contact_name(options.person)

        if options.date is None:
            log.info('Set --date switch to NOW')
            xml_object.set_date_to_now()
        else:
            now = datetime.fromtimestamp(int(options.date))
            log.info('Switch --date was set to "%s"', now)
            xml_object.set_date(now)

        if options.type == "service":
            map_alert_object_to_arguments(options, xml_object)

        if options.dumpxml is not None:
            log.debug('Dump xml to file (%s)', options.dumpxml)
            with open(options.dumpxml, 'w') as file_handle:
                file_handle.write(str(xml_object))
            return 0

        if options.noenc is True:
            data = xml_object.toxml()
        else:
            gpg = GnupgCommand(options.gnupgconfig)
            try:
                data = gpg.crypt_ascii(GNUPG_KEY, xml_object.toxml())
            except GnupgCommandException as e:
                log.error(e.message)
                return 255

        mailer = Mailer()
        mailer.server = MAIL_SERVER

        if options.targetmail is not None:
            log.debug('Override target mail address: %s', options.targetmail)
            mailer.receiver = options.targetmail
        else:
            mailer.receiver = MAIL_TARGET_ADDRESS

        if options.type == "test":
            mailer.test_mode = True

        mailer.sender = options.mail
        mailer.sender_name = options.person
        mailer.alert_type = options.type
        mailer.content = data
        mailer.send()

        time_run = time.time() - time_start

        log.info('Runtime %.4f seconds', time_run)

        if options.checkplugin is True:
            output = "%s sent %s - OK|runtime=%.4fs" % (VERSION_STRING, options.type, time_run, )
            print (output)

    except MyOptionIsMandatoryError as mandatory_error:
        print(mandatory_error.message + "\n")
        myoptions.print_usage()
        return 255
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: lib/tkmon/TKMON/Model/ThomasKrenn/Alert.php
<?php
/**
 * This file is part of TKMON
 *
 * TKMON is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * TKMON is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with TKMON.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Marius Hein <marius.hein@netways.de>
 * @copyright 2012-2013 NETWAYS GmbH <info@netways.de>
 */

namespace TKMON\Model\ThomasKrenn;

use NETWAYS\IO\Process;
use TKMON\Exception\ModelException;
use TKMON\Model\ApplicationModel;

/**
 * Abstract model for working with our DI container
 * @package TKMON\Model
 * @author Marius Hein <marius.hein@netways.de>
 */
class Alert extends ApplicationModel
{

    /**
     * Type enum for testing
     * @var string
     */
    const TYPE_TEST = 'test';

    /**
     * Type enum for heartbeat
     * @var string
     */
    const TYPE_HEARTBEAT = 'heartbeat';

    /**
     * Type enum for alerting services
     * @var string
     */
    const TYPE_SERVICE = 'service';

    /**
     * Name of preconfigures command
     * @var string
     */
    const COMMAND_NAME = 'tkalert.sh';

    /**
     * Alert Type
     *  - heartbeat
     *  - service
     *  - test
     *
     * @var string
     */
    private $type;

    /**
     * AuthKey to validate ThomasKrenn service
     * @var string
     */
    private $authKey;

    /**
     * Name of responsible person
     * @var string
     */
    private $contactName;

    /**
     * Email address of responsible person
     * @var string
     */
    private $contactMail;

    /**
     * Process object to call
     * @var Process
     */
    private $processObject;

    /**
     * Flag to indicate that object is prepared
     * @var bool
     */
    private $prepared = false;

    public function __construct(\Pimple $container)
    {
        parent::__construct($container);
        $this->processObject = $container['command']->create(self::COMMAND_NAME);
    }

    /**
     * Setter for authKey
     * @param string $authKey
     */
    public function setAuthKey($authKey)
    {
        $this->authKey = $authKey;
    }

    /**
     * Getter for authKey
     * @return string
     */
    public function getAuthKey()
    {
        return $this->authKey;
    }

    /**
     * Setter for contactMail
     * @param string $contactMail
     */
    public function setContactMail($contactMail)
    {
        $this->contactMail = $contactMail;
    }

    /**
     * Getter for contactMail
     * @return string
     */
    public function getContactMail()
    {
        return $this->contactMail;
    }

    /**
     * Setter for contactName
     * @param string $contactName
     */
    public function setContactName($contactName)
    {
        $this->contactName = $contactName;
    }

    /**
     * Getter for contactName
     * @return string
     */
    public function getContactName()
    {
        return $this->contactName;
    }

    /**
     * @param string $type
     */
    public function setType($type)
    {
        $this->type = $type;
    }

    /**
     * @return string
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * Apply configuration from ContactInfo data object
     * @param ContactInfo $contactInfo
     */
    public function configureByContactInfo(ContactInfo $contactInfo)
    {
        // Test if loaded successfully
        if (!$contactInfo->getObject()) {
            $contactInfo->load();
        }

        $this->setAuthKey($contactInfo->getAuthKey());
        $this->setContactName($contactInfo->getPerson());
        $this->setContactMail($contactInfo->getEmail());
    }

    /**
     * Creates a pre-configured process object for later use
     * @return Process
     */
    public function getProcessObject()
    {
        return $this->processObject;
    }

    /**
     * Prepare process object
     * @throws \TKMON\Exception\ModelException
     */
    public function prepare()
    {
        if (!$this->getType()) {
            throw new ModelException('$type not defined');
        }

        if (!$this->getAuthKey()) {
            throw new ModelException('$authKey not defined');
        }

        if (!$this->getContactMail()) {
            throw new ModelException('$contactMail not defined');
        }

        if (!$this->getContactName()) {
            throw new ModelException('$contactName not defined');
        }

        $this->processObject->addNamedArgument('--type', $this->getType());
        $this->processObject->addNamedArgument('--contact-person', $this->getContactName());
        $this->processObject->addNamedArgument('--contact-mail', $this->getContactMail());
        $this->processObject->addNamedArgument('--auth-key', $this->getAuthKey());

        $this->prepared = true;
    }

    /**
     * Executes the alerter
     * with previously configured parameters
     */
    public function commit()
    {
        if ($this->prepared === false) {
            $this->prepare();
        }

        $this->processObject->execute();
    }
}
